src/coreclr/src/gc/gc.cpp

```diff
@@ -24145,19 +24131,39 @@ void gc_heap::plan_phase (int condemned_gen_number)
 #ifdef MARK_LIST
             if (use_mark_list)
             {
-               while ((mark_list_next < mark_list_index) &&
-                      (*mark_list_next <= x))
-               {
-                   mark_list_next++;
-               }
-               if ((mark_list_next < mark_list_index)
+                if (current_c_gc_state != c_gc_state_marking)
+                {
+                    while ((mark_list_next < mark_list_index) &&
+                        (*mark_list_next <= x))
+                    {
+                        mark_list_next++;
+                    }
+                    x = end;
+                    if ((mark_list_next < mark_list_index)
 #ifdef MULTIPLE_HEAPS
-                   && (*mark_list_next < end) //for multiple segments
+                        && (*mark_list_next < end) //for multiple segments
 #endif //MULTIPLE_HEAPS
-                   )
-                   x = *mark_list_next;
-               else
-                   x = end;
+                        )
+                        x = *mark_list_next;
+                }
+                else
+                {
+                    assert(gc_heap::background_running_p());
+                    uint8_t* old_x = x;
+                    while ((mark_list_next < mark_list_index) &&
```

Comment:
this while loop looks identical to the one in the `(current_c_gc_state != c_gc_state_marking)` case, can they be factored out?

Reply:
yeah it'd be nicer to factor it out.

Reply:
Agree, that's the way I had done it initially - I only duplicated the while loop so I could profile all four cases separately.